File: 322_Coin_Change.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        k = len(coins)
        dp = []
        new = []
        for j in xrange(amount+1):
            new.append(j/coins[0] if j%coins[0]==0 else sys.maxint)
        dp.append(new)
        for i in xrange(1,k):
            new = []
            for j in xrange(amount+1):
                if j >= coins[i]:
                    new.append(min(dp[i-1][j], new[j-coins[i]] + 1))
                else:
                    new.append(dp[i-1][j])
            dp.append(new)
        return -1 if dp[k-1][amount] == sys.maxint else dp[k-1][amount]


class Solution:
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        maxima = float('inf')
        dp = [0] + [maxima] * amount
        for i in range(1, amount + 1):
            dp[i] = min([dp[i - c] if i - c >= 0 else maxima for c in coins]) + 1
        return [dp[-1], -1][dp[-1] == maxima]

File: 322_Coin_Change.cs
public class Solution {
    public int CoinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Array.Fill(dp, amount + 1);
        dp[0] = 0;
        
        foreach (int coin in coins) {
            for (int i = 1; i <= amount; ++i) {
                if (i >= coin) {
                    dp[i] = Math.Min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return dp[amount] == amount + 1 ? -1 : dp[amount];
    }
}
